File: nevec_rs/src/err/suggestion.rs
use super::line::Line;
use super::loc::{Loc, LocBuilder};
use super::note::Note;

#[derive(Clone, Default)]
pub struct Suggestion<'a> {
    header: Option<String>,

    loc: Option<Loc>,

    fix: Option<String>,
    explanation: Option<String>,

    original_line: Option<&'a str>,
    insert: bool,
}

impl<'a> Suggestion<'a> {
    pub fn header(mut self, header: String) -> Self {
        self.header = Some(header);
        self
    }

    pub fn at(mut self, loc: Loc) -> Self {
        self.loc = Some(loc);
        self
    }

    pub fn fix(mut self, fix: String) -> Self {
        self.fix = Some(fix);
        self
    }

    pub fn explanation(mut self, explanation: String) -> Self {
        self.explanation = Some(explanation);
        self
    }

    pub fn original_line(mut self, line: &'a str) -> Self {
        self.original_line = Some(line);
        self
    }

    pub fn insert(mut self) -> Self {
        self.insert = true;
        self
    }

    pub fn build(self) -> Result<Line, SuggestionBuildError> {
        let loc = self.loc.ok_or(SuggestionBuildError)?;
        let fix = self.fix.ok_or(SuggestionBuildError)?;
        let explanation = self.explanation.ok_or(SuggestionBuildError)?;
        let original_line = self.original_line.ok_or(SuggestionBuildError)?;

        let notes = Suggestion::make_notes(&loc, &fix, explanation);
        let line = Suggestion::modify_line(&loc, &fix, original_line, self.insert);

        Ok(Line::builder(loc)
            .header(self.header.ok_or(SuggestionBuildError)?)
            .add(notes)
            .with_line(line)
            .build())
    }

    fn make_notes(loc: &Loc, fix: &str, explanation: String) -> Vec<Note> {
        let loc = LocBuilder::from(loc).len(fix.len()).build();

        vec![Note::fix(loc, explanation)]
    }

    fn modify_line(loc: &Loc, fix: &str, original_line: &'a str, insert: bool) -> String {
        let from = loc.col - 1;
        let to = from + loc.len;

        if insert {
            format!(
                "{}{}{}",
                &original_line[..from],
                fix,
                &original_line[from..]
            )
        } else {
            format!("{}{}{}", &original_line[..from], fix, &original_line[to..])
        }
    }
}

#[derive(Debug)]
pub struct SuggestionBuildError;

File: src/err/line/Suggestion.kt
package err.line

import err.note.Note
import file.span.Loc
import file.span.LocBuilder
import file.span.indexable
import util.extension.map

/**
 * A disguised [Line] builder tailored specifically for fix suggestions.
 *
 * @property originalLine The original unmodified line.
 * @property fix The replacement you want to make at [fixLoc].
 * @property msg The message that goes under the fix [Note].
 * @property header An optional header message.
 * @property insert Whether the [fix] should be *inserted* at [fixLoc] or should be treated as a replacement at [fixLoc].
 */
class Suggestion(loc: Loc) {
    companion object {
        fun builder(loc: Loc) = Suggestion(loc)
    }

    private var fixLoc = loc

    private var originalLine: String? = null
    private var fix: String? = null
    private var msg: String? = null
    private var header: String? = null
    private var insert: Boolean = false

    fun build(): Line {
        require(originalLine != null) { "An original line must be given" }
        require(fix != null) { "A fix must be given" }
        require(msg != null) { "A message must be given" }

        val modifiedLine = modify()
        val modifiedLoc = LocBuilder.from(fixLoc).len(fix!!.length.toUInt()).build()
        val notes = listOf(Note.fix(modifiedLoc, msg!!))

        return Line.builder(modifiedLoc).withLine(modifiedLine).add(notes).header(header).build()
    }

    private fun modify(): String {
        val (from, to) = fixLoc.extremes().map(UInt::indexable)

        if (from >= originalLine!!.length) {
            return originalLine + fix!!
        }

        return if (insert) {
            originalLine!!.replaceRange(from, from, fix!!)
        } else {
            originalLine!!.replaceRange(from, to, fix!!)
        }
    }

    fun withOriginal(line: String) = apply { this.originalLine = line }

    fun withMsg(msg: String) = apply { this.msg = msg }

    fun fix(fix: String) = apply { this.fix = fix }

    fun header(msg: String?) = apply { this.header = msg }

    fun insert(insert: Boolean = true) = apply { this.insert = insert }
}